Recognize javascript functions & classes

dunkmann00 opened the pull request:
With this change, javascript functions will be recognized both when they are defined and when they are called.

tancnle reviewed (commented):
Thank you for your contribution @dunkmann00 ❤️ The PR looks great to me. 

| Before | After | 
| -- | -- |
| ![before](https://user-images.githubusercontent.com/756722/216805839-741e36a1-cad4-4707-9701-f1cf54859701.png) | ![after](https://user-images.githubusercontent.com/756722/216805843-40e2aef6-f7e7-489b-90ed-45f6f3f64ac1.png) |

I have noted `function` keyword as part of the anonymous function is no longer picked up after this MR 🤔. I can create a follow-up to address it.

dunkmann00 commented on lib/rouge/lexers/javascript.rb:
```diff
@@ -177,6 +177,13 @@ def self.id_regex
           push :expr_start
         end
 
+        rule %r/(function)((?:\s|\\\s)+)/ do
+          groups Keyword, Text
+          push :funcname
+        end
```
Yes, no problem! I think I just copied this pattern from the Python lexer, but I don't see any reason to not do it the way you suggested.

dunkmann00 commented on lib/rouge/lexers/javascript.rb:
```diff
@@ -177,6 +177,13 @@ def self.id_regex
+          push :funcname
+        end
+
+        rule %r/([a-z_]\w*)[ \t]*(?=(\(.*\)))/m, Name::Function
```
Yes...but not with `id`. The issue is the best way for `rouge` to determine if the token is a function is by checking to see if it starts with a lowercase letter. I know that this isn't actually required by `javascript`, but otherwise it would match classes as functions as well and we don't want that. I do have something that I am testing that I think will work though. I'll push it shortly to let you see what you think.

dunkmann00 commented:
@tancnle So I updated the issues you brought up in the reviews and also added classes to be recognized as well! I realized if I was going to add functions I should probably add classes. You can take a look at the regexes I used for each to see what you think. What I ended up doing is first checking for the class regex using a modified copy of the `id` regex and only matching if the first letter that appears in the name is not a lowercase character. Then for the function regex, which is checked second, I could just use the `id` regex because if it doesn't match class, it must be a function.

That might seem confusing, but hopefully it will make sense when you look at the regexes. If not, please let me know and I will try to explain it better!

| Before | After |
|--------|------|
| <img width="398" alt="image" src="https://user-images.githubusercontent.com/11963090/217150574-87bb7b54-cc55-40f3-8e87-a9ce2d908fbc.png"> | <img width="400" alt="image" src="https://user-images.githubusercontent.com/11963090/217150336-3f6c29f7-1565-482d-8408-1e4ab7084188.png"> |
| <img width="257" alt="image" src="https://user-images.githubusercontent.com/11963090/217150629-00652301-b71f-4752-a0b2-af55f1bb5e53.png"> | <img width="261" alt="image" src="https://user-images.githubusercontent.com/11963090/217150426-1918ce9d-1509-4af6-ab86-48d5384ae23b.png"> |

dunkmann00 commented:
With support for class inheritance added:

| Before | After |
|--------|------|
| <img width="271" alt="image" src="https://user-images.githubusercontent.com/11963090/217300795-4a69d92f-74f4-4c28-b4ce-5ab91a8672d7.png"> | <img width="271" alt="image" src="https://user-images.githubusercontent.com/11963090/217300612-01e46f8b-b2db-48a0-884e-2f350de34609.png"> |

tancnle commented:
Thanks for being thorough with your explanation @dunkmann00 ❤️ The PR looks great to me 🚀 💯

chethanEvEng commented:
After this PR, the `function` in following JS code is not being identified as keyword anymore. Javascript allows defining function with a space.
```
const square = function (number) {
  return number * number;
};
```
The regex needs to optionally honour the space.
